confluent_kafka/src/confluent_kafka.c

```diff
@@ -1177,6 +1216,16 @@ static void error_cb (rd_kafka_t *rk, int err, const char *reason, void *opaque)
 	CallState *cs;
 
 	cs = CallState_get(h);
+
+        /* If the client raised a fatal error we'll raise an exception
+         * rather than calling the error callback. */
+        if (err == RD_KAFKA_RESP_ERR__FATAL) {
```

Comment:
Interestingly, in C#, I think the behavior is going to be the opposite. produce methods can (and mostly will) throw for non-fatal reasons and I think it will be common for people to handle those in a generic catch-all way and typical to check for fatal errors in the error callback (though testing the exception would also be possible). i think what you have here makes sense for the python api though.